enable Pipeline to get device from model

faaany opened the pull request:
## What does this PR do?
```python
import torch 
from transformers import AutoModelForCausalLM
from transformers import AutoTokenizer, pipeline

model_id = "meta-llama/Llama-2-7b-chat-hf"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
print(model.device)
tokenizer = AutoTokenizer.from_pretrained(model_id)
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
print(pipe.model.device)
results = pipe("He's a dreadful magician and")
```
Currently, the code above will give an output of 
```bash 
cuda:0
cpu
```
But this is not OK: when users have moved the model to CUDA, Pipeline should not move the model back to CPU without showing any message. This PR makes it possible to let the model stay on its original device. Below is the results after this PR:
```bash 
cuda:0
cuda:0
```

@Narsil and @muellerzr

faaany commented:
@yao-matrix

amyeroberts reviewed (commented):
Thanks for adding this! 

Could you add a test?

muellerzr commented:
@faaany are we sure that `model.device` *is a thing* across all these frameworks?

At most I see `ModuleUtilsMixin` has `device` which is PyTorch specific (it gets added to `AutoModel`, but I'd like to verify the locations of TF and Flax backends having these capabilities to grab the model device. Otherwise we don't *really* want just `None` here IMO

faaany committed:
check model.device

faaany committed:
fix

faaany committed:
style fix

faaany commented:
> Thanks for adding this!
> 
> Could you add a test?

sure, in which test file should I put this test?

faaany commented:
> @faaany are we sure that `model.device` _is a thing_ across all these frameworks?
> 
> At most I see `ModuleUtilsMixin` has `device` which is PyTorch specific (it gets added to `AutoModel`, but I'd like to verify the locations of TF and Flax backends having these capabilities to grab the model device. Otherwise we don't _really_ want just `None` here IMO

Good point! Yes, I know that Flax model doesn't have "device". How about moving it inside `if is_torch_available() and self.framework == "pt":` ? I have updated my code.

Furthermore, I removed the `self.device is not None` check, because it will never be None. And I also added the logic that model shouldn't be moved, if the model is already on device.

amyeroberts reviewed (approved):
Thanks for updating and handling the torch case! 

Only request is to add a test. 

@muellerzr could you give a quick review as you correctly spotted and highlighted the torch vs. other frameworks case?

faaany committed:
move model device

faaany committed:
remove print

faaany committed:
add comment

faaany committed:
fix

faaany committed:
Merge branch 'huggingface:main' into device

faaany commented:
Hi @amyeroberts, sorry for the late response. We had a long holiday here in China. Unit test is added for this case. 

There are 3 possibilities for model.device:
a1. user passes `device_map`
a2. user doesn't pass `device_map`
a3. user manually moves the model to a certain device with `to(device)`

There are 2 possibilities for pipeline.device:
b1. user passes `device`
b2. user doesn't pass device 

Since a1&b1, a1&b2, a2&b1, a2&b2 are trivial, my unit tests only cover the case a3&b2, which is what my PR is trying to fix. But I also add an additional edge check a3&b1. Pls have a review, thx!

faaany committed:
add unit test

faaany committed:
optimize

faaany committed:
change test names and add more cases

amyeroberts commented on tests/pipelines/test_pipelines_common.py:
```diff
@@ -519,6 +520,52 @@ def test_pipeline_negative_device(self):
+            "hf-internal-testing/tiny-random-bert", torch_dtype=torch.float16
+        ).to(model_device)
+        target_device = f"{torch_device}:0"
+        assert model_device != target_device
+        pipe = pipeline("text-generation", model=model, device=target_device, tokenizer=tokenizer)
+        assert pipe.model.device == torch.device(target_device)
```
nit 

```suggestion
        self.assertNotEqual(model_device, target_device)
        pipe = pipeline("text-generation", model=model, device=target_device, tokenizer=tokenizer)
        self.assertEqual(pipe.model.device, torch.device(target_device))
```

amyeroberts reviewed (approved):
Looks great - thanks for adding the tests and the explanation! 

cc @muellerzr For a final double check to make sure this makes sense with accelerate

muellerzr reviewed (approved):
Much better, thanks! Agreed post Amy's nit :)

faaany committed:
Update tests/pipelines/test_pipelines_common.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faaany commented:
Thanks for the review! @amyeroberts @muellerzr
